resources/templates/md/posts/2016-04-04-state.md

```diff
@@ -0,0 +1,370 @@
+In this function we bind the new value for `marshmallows-in-the-bag` (difference between `marshmallows-in-the-bag` and `marshmallows-eaten`).
+
+Now we call `number-of-marshmallows-left` with 10 as an argument (as we know, Greg ate 10 marshmallows). 
+
+```Clojure
+(number-of-marshmallows-left 10)
+
+;;=> Marshmallows: 20
+```
+
+To show the difference we'll do the same using `let block` (`lexical scope`) instead of `binding` (`dynamic scope`). 
+
+```Clojure
+(def ^:dynamic *marshmallows-in-the-bag* 30)
+
+(defn marshmallows-left []
+  (println (str "Marshmallows: " *marshmallows-in-the-bag*)))
+
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (let [*marshmallows-in-the-bag* (- *marshmallows-in-the-bag* marshmallows-eaten)]
+    (marshmallows-left)))
+```
+
+And if we call `numbers-of-marshmallows-left` with `10` as an argument, the result will be different as before.
+
+```Clojure
+(number-of-marshmallows-left 10)
+
+;;=> Marshmallows: 30
+```
+
+That's because we defined `marshmallows-in-the-bag` in the `let` code block and as we know the new value of the `marshmallows-in-the-bag` 
+is valid only within it's lexical scope. So for the `marshmallows-left` function its the value of the `marshmallows-in-the bag` still `30`. 
+
+### Atoms
+
+Clojure's `atom` reference type is another way how we can manage state. 
+They are updated synchronously and they are ideal for managing the state of independend identities.
+
+We will use the same example as we used in `vars`. 
+In this case we will define `marshmallows-in-the-bag` as an atom.
+
+```clojure
+(def marshmallows-in-the-bag (atom 30))
+```
+
+To get the current state of the atom, we need to dereference it using `clojure.core/deref` or `@` reader macro.
+
+```clojure
+@marshmallows-in-the-bag
+
+;;=> 30
+```
+
+To update the atom value (change what the atom refers to), we will use `swap!`.
+Remember that the value of the atom is not changed, the reference is updated and assigned the new value.
+
+How the `swap` behaves:
+
+1. Read the atom's current state.
+2. Compare the current state of the `atom` and the state where the `swap` startet with transaction.
+3. If there's difference, make the transaction retry.
+4. If there's not difference, update the state.
+
+```clojure
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (swap! marshmallows-in-the-bag - marshmallows-eaten))
+```
+We created a `number-of-marshmalloes-left` function, which takes one argument `marshmallows-eaten`.
+We passed `swap!` a function which will be called with `current-state` as an first argument. 
+This function subtracts `marshmallows-eaten` from the current state.
+
+```clojure
+(number-of-marshmallows-left 10)
+
+;;=> 20
+```
+We can dereference the atom, and we will see that the `marshmallows-in-the-bag` value is updated
+
+```clojure
+@marshmallows-in-the-bag
+
+;;=> 20
+```
+
+## Agents
+
+`Agents` are similiar to `Atoms`, but in contrast with `atoms`, they are updated asynchronously.
```

Comment:
Here, as well as in all other parts, I would use the lower case version: `atom`s, `agent`s,..